### DISCUSSION TITLE
Granian Container config

### DISCUSSION THREAD

**Discussion by b3n4kh** [emmett-framework/granian#271] [Q&A] [answered]
I am trying to migrate my existing gunicorn config to granian.
In there I have a gunicorn.conf where I do something like this:

```
workers_per_core = int(os.getenv("WORKERS_PER_CORE", "2"))
workers = int(workers_per_core * multiprocessing.cpu_count())
```

Now I would like to move that into my Containerfile where I got:

```
ENV GRANIAN_PORT=8080
ENV GRANIAN_HOST=0.0.0.0
ENV GRANIAN_INTERFACE=wsgi
ENV GRANIAN_APP=run

# GRANIAN_WORKERS=
# GRANIAN_THREADS=
# GRANIAN_BLOCKING_THREADS=

CMD ["granian", "run"]
```

Any ideas as to get a similar setup based on the number of cores? Should I approach this differently and write some kind of wrapper script?

**Comment by gi0baro** [comment, marked as answer]
@b3n4kh if you want to run those computation, you probably want to launch granian programmatically using a Python file (see for example https://github.com/emmett-framework/granian/discussions/254).

Given that you run the application in a container, I would keep processes/threads number low and use containers to scale your app.
This is generally what I suggest when running services using containers (eg: 1 proc and 2 blocking threads).

Also, if you keep the cli usage, the `CMD` line looks wrong, as there is no `granian run` command, but just `granian`.

**Reply by b3n4kh** [reply]
Thanks for the reply! I think I will go with a similar script! Sorry for the confusion "run" is just the name of my "app". But by starting it via the python interface, I can diretcly pass it which is much nicer! Thx!

**Reply by b3n4kh** [reply]
I got it running, I needed the `target_loader` config as described here: https://github.com/emmett-framework/granian/issues/35#issuecomment-1396232053

But with this I got it running "as expected".

**Comment by b3n4kh** [comment]
Is there a feature similar to `gunicorns` server hooks? Especially the [post_fork](https://docs.gunicorn.org/en/stable/settings.html#server-hooks)?

Or any other way to identify workers?

**Reply by gi0baro** [reply]
No, but you can write your own `spawn_target` method (https://github.com/emmett-framework/granian/blob/v1.2.3/granian/server.py#L478) which might just be a wrapper of the default ones. Such method will receive among other stuff the `worker_id`, if that's just what you need.
